### ISSUE TITLE
Vector related issues

### ISSUE THREAD

**Issue by mcovert** [jpmml/jpmml-sparkml#108]
Has anyone been able to successfully use this library with Spark ML 2+? Specifically, I ask this question due to the Vector IllegalArgumentException when processing algorithm transformers. I have looked through most of the classifiers and regressors and it appears that every algorithm expects a Vector to make its prediction. Given that Vector is not a supported type, it is not clear to me how this can ever be made to work.

I have tried every suggested workaround that I can find.  I inserted a VectorIndexer after my VectorAssembler, but its output is also a Vector.  I have tried to manipulate the StructType/StructField schema objects. I have even tried to set a DecisionTreeModel numFeatures to 1, but nothing works. So far, all that I can make work is to produce PMML for the feature transformers, but any time I add an algorithmModel, it seems to fail with either the IllegalArgumentException or a "expected nn features but got 1".

I have examined the schema object (StructType) that I am passing to JPMML build and in it I can see the "features" column VectorAssembler and its mapping to prior categorical fields. This works because I am using the schema of the DataFrame that trained and tested the algorithm. What I assume is the the JPMML converter is "expecting" the fields that are hidden in the "features" column.  Is there any trick to editing the features column StructField metadata to make JPMML-sparkml extract the "sub-features"?

I have spent about 2 weeks on this and would really like to get it to work, but I am just baffled. Can anyone help? Does anyone have a working code example?

**Comment by vruusmann** [comment]
Closing as exact duplicate of https://github.com/jpmml/jpmml-sparkml/issues/26

> Has anyone been able to successfully use this library with Spark ML 2+?

See https://github.com/jpmml/jpmml-sparkml#library-1

The JPMML-SparkML library supports Apache Spark ML 1.X, 2.X and 3.X.

> Given that Vector is not a supported type

JPMML-SparkML assumes **structured data**, which means that all the columns entering the pipeline must be of some scalar data type (integer, float, boolean string etc). Once the structured data is inside a pipeline, then it can be assembled into vectors, sliced, indexed etc.

A vector column in isolation is **unstructured data**. If you want to start your pipeline with a vector column, then you'd need to write a custom transformer which kind of explains the meaning of each vector element. AFAIK, there's no "vector explained" transformation in Apache Spark ML core library.

**Comment by mcovert** [comment]
I can appreciate you closing this issue, but the example that you provide uses the Iris.csv dataset and RFormula. RFormula outputs a label (double) and features (a Vector). I have a more complex application that has 15 features, a long sequence of transformations for many of them, and the final result is assembled into a vector (and indexed), so essentially the tail of my encoder pipeline ends up exactly like your example. 

So, I do not start my pipeline with a vector column. I start it with 2 doubles and 13 string columns. The doubles are scaled and bucketized and the strings are string indexed. After each is transformed they are all doubles, categorical. They must then be assembled into something that a classifier can handle (i.e. a column named "features" which is of type vector). So after assembly, it is this column that is passed to three algorithms that end the pipeline. In the schema StructType that is passed to PMML build, , I see the features column has metadata with all of its assembled parts there and each part is specified as as categorical). 

So, why is it that the Iris version works but mine does not? And, is there a working example that more closely resembles mine? I am willing to try the RFormula approach rather than VectorAssembler if it does something that will work on the JPMML side.

One other thing. I am absolutely not trying to be argumentative or in any way disparaging. I have looked at your code very closely and I think it is elegant and I also think that I understand much of it (and have traced it through). I really just need to either find a way to do this or I must move on without including PMML output in this release (which I really do not want to do).

**Comment by vruusmann** [comment]
> RFormula outputs a label (double) and features (a Vector).

RFormula takes many scalar columns as input and builds a single vector column out of them. The JPMML-SparkML library is able to keep track of this vector assembly process.

> I have a more complex application that has 15 features

The number of input columns is not limited in any way. Just keep them as 15 separate scalar columns (when they enter the pipeline).

>  I start it with 2 doubles and 13 string columns.

Then where does this vector column come from? Perhaps you have a _vector of two doubles_, and 13 strings?

> So, why is it that the Iris version works but mine does not? 

I didn't compile your pipeline. In fact, I haven't even seen it.

Can you isolate your use case into an iris-style example (ie. the minimum number of code lines, runs fine on Apache Spark, but fails to convert with JPMML-SparkML)?

> And, is there a working example that more closely resembles mine?

See https://github.com/jpmml/jpmml-sparkml#documentation

Then jump to https://openscoring.io/blog and see Apache Spark ML-related posts. For example, https://openscoring.io/blog/2020/01/19/converting_logistic_regression_pmml/

**Comment by mcovert** [comment]
The vector column is produced by the VectorAssembler. There is no vector data in any input, only doubles and strings. Here is a description of the pipeline:

Input Column                  Transformer Chain              Output Column
------------------------      -----------------------         -------------------    
f1 - f2 (both Doubles)  ->  scaler -> bucketizer   -> f1_index - f2_index
f3 - f15 (all Strings)      ->  StringIndexer             ->  f3_index - f15_index   

Then all *_index columns (now, all categorical Doubles) are assembled into a single features column (a Vector),
f1_index to f15_index -> VectorAssembler -> features_raw  -> VectorIndexer -> features -> DecisionTreeClassifier

There is a chain of 3 algorithms at the end one for each target that I need. one is a classifer and two are regressors. Each is trained and the model is tested. Each model is then converted to PMML and written to a database.

That's really all there is to it. I just looked at the code for RFormula and it appends the same metadata that VectorIndexer does.

I have read all of the documentation that you cite and have seen all of the code as well. None really address my problem though.  Later tonight I will build a small subset of code and post the results for you to see. I will include the StructType JSON also.

Again, thank you very much for taking the time to address this.

**Comment by vruusmann** [comment]
> Here is a description of the pipeline:

I'd like to have it executable (script plus a CSV file with a couple of data records), because I want to see the exception happening!

Could you at least paste a full exception stack trace here? It shows exactly which pipeline stage, and in which context, balks at a vector.

**Comment by mcovert** [comment]
I will. I can tell you that all encoder stages work just fine and it is the DecisionTreeClassifierModel stage that fails. But I'll have everything packaged up soon and get it to you.

**Comment by mcovert** [comment]
I think I have found my problem and it is indeed on my side. 

We manually create a PipelineModel because many of our stages are "pre-fitted" (they are stored in a library). In general, we do not use Pipelines because in addition to batch processing, the models are served up from an Akka-Http REST server, and the DataFrame API that they use is too slow here. Unfortunately, Databricks has decided to mark a lot of needed methods as protected or private (although they are opening them up slowly), so the process is more complicated that it needs to be.

During model train and test, everything works as expected. However, during the creation of the Pipeline model required by JPMML-Sparkml, the insertion of the VectorAssembler is handled incorrectly. I looked through the Pipeline.fit() method and found the mistake. After this correction everything seems to work. I will be doing the full integration and QA today. If you don't mind, I may post some information about all of this because I believe that several others have encountered a different but similar set of issues.

My confusion stemmed both from my interpretation of the documentation, and also from misinterpreting several of the prior threads. When it was stated that JPMML did not support Vectors, I read this totally the wrong way. Based on this thread, I now fully understand the issue with Vectors at the head of the pipeline. Although metadata could be manually inserted, Databricks makes this difficult and it isn't well documented. I will add that my experience has been to be wary of using some Spark low level functionality because it can be very unstable over time. Also, reading their ML source code makes my head hurt! :-) 

I am attaching some code so that you see what I did, but I will not forward any log of stack traces because it is irrelevant.

I would like to thank you very much for your help here, and I would like to apologize for asking you to do this. It did help me immensely though and I very much appreciate your support.

[TestJPMML.scala.txt](https://github.com/jpmml/jpmml-sparkml/files/6142110/TestJPMML.scala.txt)
